chain/src/txhashset/bitmap_accumulator.rs

```diff
@@ -0,0 +1,241 @@
+///
+#[derive(Clone)]
+pub struct BitmapAccumulator {
+	backend: VecBackend<BitmapChunk>,
+}
+
+impl BitmapAccumulator {
+	/// Crate a new empty bitmap accumulator.
+	pub fn new() -> BitmapAccumulator {
+		BitmapAccumulator {
+			backend: VecBackend::new_hash_only(),
+		}
+	}
+
+	/// Initialize a bitmap accumulator given the provided idx iterator.
+	pub fn init<T: IntoIterator<Item = u64>>(&mut self, idx: T) -> Result<(), Error> {
+		self.apply_from(idx, 0)
+	}
+
+	/// Apply the provided idx iterator starting at the provided chunk idx.
+	fn apply_from<T>(&mut self, idx: T, from_chunk_idx: u64) -> Result<(), Error>
+	where
+		T: IntoIterator<Item = u64>,
+	{
+		let now = Instant::now();
+
+		let mut chunk_idx = from_chunk_idx;
+		let mut chunk = BitmapChunk::new();
+
+		let mut leaves = idx.into_iter().peekable();
+		while let Some(x) = leaves.peek() {
+			if *x < chunk_idx * 1024 {
+				// skip until we reach our first chunk
+				leaves.next();
+			} else if *x < (chunk_idx + 1) * 1024 {
+				let idx = leaves.next().expect("next after peek");
+				chunk.set(idx % 1024, true);
+			} else {
+				self.append_chunk(chunk)?;
+				chunk_idx += 1;
+				chunk = BitmapChunk::new();
+			}
+		}
+		if chunk.any() {
+			self.append_chunk(chunk)?;
+		}
+		debug!(
+			"applied {} chunks from idx {} to idx {} ({}ms)",
+			1 + chunk_idx - from_chunk_idx,
+			from_chunk_idx,
+			chunk_idx,
+			now.elapsed().as_millis(),
+		);
+		Ok(())
+	}
+
+	/// Apply updates to the bitmap accumulator given an iterator of invalidated idx and
+	/// an additional iterator of new idx to be added.
+	/// We determine the existing chunks to be rebuilt given the invalidated idx.
+	/// We then add new idx, extending the accumulator with new chunk(s) as necessary.
+	pub fn apply<T, U>(&mut self, invalidated_idx: T, new_idx: U) -> Result<(), Error>
+	where
+		T: IntoIterator<Item = u64>,
+		U: IntoIterator<Item = u64>,
+	{
+		// Determine the earliest chunk by looking at the min invalidated idx (assume sorted).
+		// Rewind prior to this and reapply new_idx.
+		// Note: We rebuild everything after rewind point but much of the bitmap may be
+		// unchanged. This can be further optimized by only rebuilding necessary chunks and
+		// rehashing.
+		if let Some(first_chunk_idx) = invalidated_idx.into_iter().next().map(|x| x / 1024) {
+			self.rewind_prior(first_chunk_idx)?;
+			self.pad_left(first_chunk_idx)?;
+			self.apply_from(new_idx, first_chunk_idx)?;
+		}
+
+		// Trim any "empty" chunks from the right to keep everything consistent.
+		self.trim_right()?;
+
+		Ok(())
+	}
+
+	/// Given the provided chunk idx rewind the bitmap accumulator to the end of the
+	/// previous chunk ready for the updated chunk to be appended.
+	fn rewind_prior(&mut self, chunk_idx: u64) -> Result<(), Error> {
+		let last_pos = self.backend.size();
+		let mut pmmr = PMMR::at(&mut self.backend, last_pos);
+		let chunk_pos = pmmr::insertion_to_pmmr_index(chunk_idx + 1);
```

Comment:
doesn't a chunk index need to be multiplies by 1024 to get an MMR insertion index?

Reply:
Not here no. We are rewinding the MMR that backs the "accumulator" so each pos here represents a 1024 bit chunk.

Reply:
ah, yes, you're right. this MMR is the UBMT...